### ISSUE TITLE
Unable to get messages from beginning for new topics with Rebalance_cb

### ISSUE THREAD

**Issue by dshivashankar1994** [edenhill/librdkafka#2938]
Description
===========

when i subscribe to existing + non-existing topics (A,B) with the offset as "beginning", the topic a receives all the messages in topic "A". Now when "B" gets created and the producer has produced 5 messages after which the meta-data-refresh happened. Now rebalance cb is called and the producer has produced 5 more messages in topic "B". The consumer receives only those messages that were published after the rebalance cb ie, last 5 messages.

How to reproduce
================
```
 LOG_INFORMATION(myLogger, "Rebalance callback called with " << partitions.size() << " partition(s).");
    if (err == RdKafka::ERR__ASSIGN_PARTITIONS) {
        if (partitions.size() == 0) {
            throw KafkaException("Partitions not assigned to consumer");
        }
        for (auto& partition : partitions) {
            LOG_INFORMATION(myLogger, "Rebalance callback called with topic" << partition->topic());
            if (offsetSeeked.count(partition->topic()) == 0) {
                partition->set_offset(RdKafka::Topic::OFFSET_BEGINNING);
                offsetSeeked.insert(partition->topic());
            }
        }
        consumer->assign(partitions);
    }
    else if (err == RdKafka::ERR__REVOKE_PARTITIONS) {
        consumer->unassign();
    }
    else {
        LOG_ERROR(myLogger, "Rebalancing error: " << RdKafka::err2str(err));
    }
}
```

- Consumer subscribes to topics A and B (A is existing and B doesn't)
- In rebalance_cb, I am setting the offset as begnning and the topic A receives all the messages from beginning
- Now producer has created the topic and produced 5 messages
- After 5 minutes, rebalance_cb is called and the offset is set to beginning and then assigned.
- Now when I describe the consumer group, I see that the CURRENT-OFFSET as '-' for topic B and LOG-END-OFFSET is 5
- Now producer produced 5 more messages
- But the consumer has received only last 5 messages on topic B (ie, only those messages after which the topic is created)
- Now when I describe the consumer group, I see that the CURRENT-OFFSET as 10 for topic B and LOG-END-OFFSET is also 10

(I maintain offsetSeeked to make sure the I don't reset the offset for the topics that are already assigned)

**IMPORTANT**: Always try to reproduce the issue on the latest released version (see https://github.com/edenhill/librdkafka/releases), if it can't be reproduced on the latest version the issue has been fixed.

Checklist
=========

**IMPORTANT**: We will close issues where the checklist has not been completed.

Please provide the following information:

 - [x] librdkafka version (release number or git tag): `<REPLACE with e.g., v0.10.5 or a git sha. NOT "latest" or "current">`
 - [ ] Apache Kafka version: `<REPLACE with e.g., 0.10.2.3>`
 - [ ] librdkafka client configuration: `<REPLACE with e.g., message.timeout.ms=123, auto.reset.offset=earliest, ..>`
 - [ ] Operating system: `<REPLACE with e.g., Centos 5 (x64)>`
 - [ ] Provide logs (with `debug=..` as necessary) from librdkafka
 - [ ] Provide broker log excerpts
 - [ ] Critical issue

**Comment by dshivashankar1994** [comment]
I modified the code to something like below

```
        for (auto& partition : partitions) {
            if (myoffsetSeeked.count(partition->topic()) == 0) {
                switch (myOffsetPosition) {
                case OffsetPosition::beginning:
                    partition->set_offset(RdKafka::Topic::OFFSET_BEGINNING);
                    break;
                case OffsetPosition::end:
                    partition->set_offset(RdKafka::Topic::OFFSET_END);
                    break;
                case OffsetPosition::stored:
                default:
                    partition->set_offset(RdKafka::Topic::OFFSET_STORED);
                    break;
                }
                myoffsetSeeked.insert(partition->topic());
            }
            else {
                partition->set_offset(RdKafka::Topic::OFFSET_STORED);
            }
        }
```

Now when I have the topic.metadata.refresh.interval.ms as 10 seconds, things are working fine. But the same is not working if I have it as 2 minutes. Am I missing anything here ?

**Comment by dshivashankar1994** [comment]
Logs:
```
[135620 (140199166191360) - INFO 20200622 20:57:49.269 IST KafkaReader.C#rebalance_cb:79] Rebalance callback called with 2 partition(s).
[135620 (140199166191360) - ERROR 20200622 20:57:49.269 IST KafkaReader.C#rebalance_cb:109] Rebalancing part: fps.dhamodha.c.1  -1000
[135620 (140199166191360) - ERROR 20200622 20:57:49.269 IST KafkaReader.C#rebalance_cb:109] Rebalancing part: fps.dhamodha.g.4  -2
[135620 (140199441028864) - DEBUG 20200622 20:57:49.269 IST KafkaReader.C#event_cb:144] 7, facility DESP, str [thrd:main]: Setting topic fps.dhamodha.c.1 [0] par
tition as desired
[135620 (140199441028864) - DEBUG 20200622 20:57:49.270 IST KafkaReader.C#event_cb:144] 7, facility TOPIC, str [thrd:main]: New local topic: fps.dhamodha.g.4
[135620 (140199441028864) - DEBUG 20200622 20:57:49.270 IST KafkaReader.C#event_cb:144] 7, facility TOPPARNEW, str [thrd:main]: NEW fps.dhamodha.g.4 [-1] 0x7f829
8001940 (at rd_kafka_topic_new0:440)
[135620 (140199441028864) - DEBUG 20200622 20:57:49.270 IST KafkaReader.C#event_cb:144] 7, facility STATE, str [thrd:main]: Topic fps.dhamodha.g.4 changed state 
unknown -> exists
[135620 (140199441028864) - DEBUG 20200622 20:57:49.270 IST KafkaReader.C#event_cb:144] 7, facility PARTCNT, str [thrd:main]: Topic fps.dhamodha.g.4 partition co
unt changed from 0 to 1
[135620 (140199441028864) - DEBUG 20200622 20:57:49.270 IST KafkaReader.C#event_cb:144] 7, facility TOPPARNEW, str [thrd:main]: NEW fps.dhamodha.g.4 [0] 0x7f8298
001ea0 (at rd_kafka_topic_partition_cnt_update:749)
[135620 (140199441028864) - DEBUG 20200622 20:57:49.270 IST KafkaReader.C#event_cb:144] 7, facility METADATA, str [thrd:main]:   Topic fps.dhamodha.g.4 partition
 0 Leader 5
[135620 (140199441028864) - DEBUG 20200622 20:57:49.270 IST KafkaReader.C#event_cb:144] 7, facility BRKDELGT, str [thrd:main]: fps.dhamodha.g.4 [0]: delegate to 
broker sasl_plaintext://u-kafkatst-kafkadev-5.sd.deshaw.com:9092/5 (rktp 0x7f8298001ea0, term 0, ref 2)
[135620 (140199441028864) - DEBUG 20200622 20:57:49.270 IST KafkaReader.C#event_cb:144] 7, facility BRKDELGT, str [thrd:main]: fps.dhamodha.g.4 [0]: delegating t
o broker sasl_plaintext://u-kafkatst-kafkadev-5.sd.deshaw.com:9092/5 for partition with 0 messages (0 bytes) queued
[135620 (140199441028864) - DEBUG 20200622 20:57:49.270 IST KafkaReader.C#event_cb:144] 7, facility BRKMIGR, str [thrd:main]: Migrating topic fps.dhamodha.g.4 [0
] 0x7f8298001ea0 from (none) to sasl_plaintext://u-kafkatst-kafkadev-5.sd.deshaw.com:9092/5 (sending PARTITION_JOIN to sasl_plaintext://u-kafkatst-kafkadev-5.sd.
deshaw.com:9092/5)
[135620 (140199441028864) - DEBUG 20200622 20:57:49.270 IST KafkaReader.C#event_cb:144] 7, facility DESP, str [thrd:main]: Setting topic fps.dhamodha.g.4 [0] par
tition as desired
[135620 (140199441028864) - DEBUG 20200622 20:57:49.270 IST KafkaReader.C#event_cb:144] 7, facility FETCHSTART, str [thrd:main]: List with 2 partition(s):
[135620 (140199441028864) - DEBUG 20200622 20:57:49.270 IST KafkaReader.C#event_cb:144] 7, facility FETCHSTART, str [thrd:main]:  fps.dhamodha.c.1 [0] offset STO
RED
[135620 (140199441028864) - DEBUG 20200622 20:57:49.270 IST KafkaReader.C#event_cb:144] 7, facility FETCHSTART, str [thrd:main]:  fps.dhamodha.g.4 [0] offset BEG
INNING
[135620 (140199441028864) - DEBUG 20200622 20:57:49.270 IST KafkaReader.C#event_cb:144] 7, facility OFFSET, str [thrd:main]: GroupCoordinator/3: OffsetFetchReque
st: skipping fps.dhamodha.g.4 [0] with valid offset BEGINNING
[135620 (140199441028864) - DEBUG 20200622 20:57:49.270 IST KafkaReader.C#event_cb:144] 7, facility OFFSET, str [thrd:main]: GroupCoordinator/3: OffsetFetchReque
st(v1) for 1/2 partition(s)
[135620 (140199182976768) - DEBUG 20200622 20:57:49.270 IST KafkaReader.C#event_cb:144] 7, facility TOPBRK, str [thrd:sasl_plaintext://u-kafkatst-kafkadev-5.sd.d
eshaw.com:9092/boots]: sasl_plaintext://u-kafkatst-kafkadev-5.sd.deshaw.com:9092/5: Topic fps.dhamodha.g.4 [0]: joining broker (rktp 0x7f8298001ea0, 0 message(s)
 queued)
[135620 (140199182976768) - DEBUG 20200622 20:57:49.270 IST KafkaReader.C#event_cb:144] 7, facility BROADCAST, str [thrd:sasl_plaintext://u-kafkatst-kafkadev-5.s
d.deshaw.com:9092/boots]: Broadcasting state change
[135620 (140199441028864) - DEBUG 20200622 20:57:49.470 IST KafkaReader.C#event_cb:144] 7, facility OFFSET, str [thrd:main]: Topic fps.dhamodha.c.1 [0]: stored o
ffset -1001, committed offset 5: not including in commit
[135620 (140199441028864) - DEBUG 20200622 20:57:49.470 IST KafkaReader.C#event_cb:144] 7, facility OFFSET, str [thrd:main]: Topic fps.dhamodha.g.4 [0]: stored o
ffset -1001, committed offset -1001: not including in commit
[135620 (140199441028864) - DEBUG 20200622 20:57:49.475 IST KafkaReader.C#event_cb:144] 7, facility OFFSETFETCH, str [thrd:main]: List with 2 partition(s):
[135620 (140199441028864) - DEBUG 20200622 20:57:49.475 IST KafkaReader.C#event_cb:144] 7, facility OFFSETFETCH, str [thrd:main]:  fps.dhamodha.c.1 [0] offset ST
ORED
[135620 (140199441028864) - DEBUG 20200622 20:57:49.475 IST KafkaReader.C#event_cb:144] 7, facility OFFSETFETCH, str [thrd:main]:  fps.dhamodha.g.4 [0] offset BE
GINNING
[135620 (140199441028864) - DEBUG 20200622 20:57:49.475 IST KafkaReader.C#event_cb:144] 7, facility OFFSET, str [thrd:main]: Topic fps.dhamodha.c.1 [0]: setting 
default offset INVALID
[135620 (140199441028864) - DEBUG 20200622 20:57:49.475 IST KafkaReader.C#event_cb:144] 7, facility OFFSET, str [thrd:main]: Topic fps.dhamodha.g.4 [0]: setting 
default offset INVALID
[135620 (140199441028864) - DEBUG 20200622 20:57:49.475 IST KafkaReader.C#event_cb:144] 7, facility OFFSETFETCH, str [thrd:main]: GroupCoordinator/3: OffsetFetch
Response: fps.dhamodha.c.1 [0] offset 5, metadata 0 byte(s)
[135620 (140199441028864) - DEBUG 20200622 20:57:49.475 IST KafkaReader.C#event_cb:144] 7, facility OFFFETCH, str [thrd:main]: GroupCoordinator/3: OffsetFetch fo
r 1/2 partition(s) returned Success
[135620 (140199441028864) - DEBUG 20200622 20:57:49.475 IST KafkaReader.C#event_cb:144] 7, facility FETCHSTART, str [thrd:main]: List with 2 partition(s):

[135620 (140199441028864) - DEBUG 20200622 20:57:49.475 IST KafkaReader.C#event_cb:144] 7, facility FETCHSTART, str [thrd:main]:  fps.dhamodha.c.1 [0] offset 5
[135620 (140199441028864) - DEBUG 20200622 20:57:49.475 IST KafkaReader.C#event_cb:144] 7, facility FETCHSTART, str [thrd:main]:  fps.dhamodha.g.4 [0] offset INVALID
[135620 (140199441028864) - DEBUG 20200622 20:57:49.475 IST KafkaReader.C#event_cb:144] 7, facility BARRIER, str [thrd:main]: fps.dhamodha.c.1 [0]: rd_kafka_toppar_op_fetch_start:2297: new version barrier v6
[135620 (140199441028864) - DEBUG 20200622 20:57:49.475 IST KafkaReader.C#event_cb:144] 7, facility CONSUMER, str [thrd:main]: Start consuming fps.dhamodha.c.1 [0] at offset 5 (v6)
[135620 (140199441028864) - DEBUG 20200622 20:57:49.475 IST KafkaReader.C#event_cb:144] 7, facility BARRIER, str [thrd:main]: fps.dhamodha.g.4 [0]: rd_kafka_toppar_op_fetch_start:2297: new version barrier v2
[135620 (140199441028864) - DEBUG 20200622 20:57:49.475 IST KafkaReader.C#event_cb:144] 7, facility CONSUMER, str [thrd:main]: Start consuming fps.dhamodha.g.4 [0] at offset INVALID (v2)
[135620 (140199441028864) - DEBUG 20200622 20:57:49.475 IST KafkaReader.C#event_cb:144] 7, facility OP, str [thrd:main]: fps.dhamodha.c.1 [0] received op FETCH_START (v6) in fetch-state stopped (opv5)
[135620 (140199441028864) - DEBUG 20200622 20:57:49.475 IST KafkaReader.C#event_cb:144] 7, facility FETCH, str [thrd:main]: Start fetch for fps.dhamodha.c.1 [0] in state stopped at offset 5 (v6)
[135620 (140199441028864) - DEBUG 20200622 20:57:49.475 IST KafkaReader.C#event_cb:144] 7, facility PARTSTATE, str [thrd:main]: Partition fps.dhamodha.c.1 [0] changed fetch state stopped -> active
[135620 (140199441028864) - DEBUG 20200622 20:57:49.475 IST KafkaReader.C#event_cb:144] 7, facility FETCH, str [thrd:main]: Partition fps.dhamodha.c.1 [0] start fetching at offset 5
[135620 (140199441028864) - DEBUG 20200622 20:57:49.475 IST KafkaReader.C#event_cb:144] 7, facility OP, str [thrd:main]: fps.dhamodha.g.4 [0] received op FETCH_START (v2) in fetch-state none (opv1)
[135620 (140199441028864) - DEBUG 20200622 20:57:49.475 IST KafkaReader.C#event_cb:144] 7, facility FETCH, str [thrd:main]: Start fetch for fps.dhamodha.g.4 [0] in state none at offset INVALID (v2)
[135620 (140199441028864) - DEBUG 20200622 20:57:49.475 IST KafkaReader.C#event_cb:144] 7, facility PARTSTATE, str [thrd:main]: Partition fps.dhamodha.g.4 [0] changed fetch state none -> offset-query
[135620 (140199441028864) - DEBUG 20200622 20:57:49.475 IST KafkaReader.C#event_cb:144] 7, facility OFFSET, str [thrd:main]: fps.dhamodha.g.4 [0]: offset reset (at offset INVALID) to END: no previously committed offset available: Local: No offset stored
[135620 (140199441028864) - DEBUG 20200622 20:57:49.475 IST KafkaReader.C#event_cb:144] 7, facility OFFSET, str [thrd:main]: fps.dhamodha.g.4 [0]: backoff: (re)starting offset query timer for offset END

```

**Comment by dshivashankar1994** [comment]
Any updates on this @edenhill & @mhowlett ?
Is there anything else I can provide to you that would help?

**Comment by edenhill** [comment]
This looks like a bug where if an assignment is set with assign() and there is a mix of partitions that are to use committed offsets and logical offsets (such as beginning) the logical offsets are lost.

Will try to reproduce, thanks for reporting.

As a workaround you can set `auto.offset.reset=earliest`.

**Comment by dshivashankar1994** [comment]
hi, any update on this ?

**Comment by dshivashankar1994** [comment]
Ping for the above

**Comment by dniceswa** [comment]
I've seen this as well. In rebalance_cb with two TopicPartitions, if I set one offset to OFFSET_BEGINNING and do nothing with the other offset, the OFFSET_BEGINNING TopicPartition turns into "INVALID" in the debug logs and then falls back on the auto.offset.reset default. If I set the second offset to OFFSET_END instead, both TopicPartitions start reading from the expected offset and neither is reported as "INVALID".
